src/Product.php

```diff
@@ -0,0 +1,176 @@
+    echo '<div class="options_group">';
+    echo '<h3 style="padding-left: 0 !important;">' . $sectionTitle . '</h3>';
+    foreach ($fields as $field) {
+      $fieldId = sprintf('_%s_%s_%s', Plugin::PREFIX, $fieldsIdPrefix, $field['id']);
+      woocommerce_wp_text_input([
+        'wrapper_class' => 'form-row',
+        'id' => $fieldId . '[' . $loop . ']',
+        'label' => $field['name'],
+        'value' => get_post_meta($variationId, $fieldId, TRUE),
+      ]);
+    }
+    echo '</div>';
+  }
+
+  /**
+   * Processes simple product meta.
+   *
+   * @implements woocommerce_process_product_meta
+   */
+  public static function woocommerce_process_product_meta($postId) {
+    if ($warehouses = Config::get('warehouses')) {
+      static::saveSimpleProductCustomFields($warehouses, 'warehouse', $postId);
+    }
+
+    if ($showrooms = Config::get('showrooms')) {
+      static::saveSimpleProductCustomFields($showrooms, 'showroom', $postId);
+    }
+  }
+
+  /**
+   * Saves custom fields for simple products.
+   *
+   * @param array $fields
+   *   Fields to render.
+   * @param string $fieldsIdPrefix
+   *   Prefix used to identify the fields.
+   * @param string $postId
+   *   The WordPress Post ID.
+   */
+  protected static function saveSimpleProductCustomFields($fields, $fieldsIdPrefix, $postId) {
+    foreach ($fields as $field) {
+      $field = sprintf('_%s_%s_%s', Plugin::PREFIX, $fieldsIdPrefix, $field['id']);
+      if (isset($_POST[$field])) {
+        if (!is_array($_POST[$field]) && $_POST[$field]) {
+          update_post_meta($postId, $field, $_POST[$field]);
+        }
+        else {
+          delete_post_meta($postId, $field);
+        }
+      }
+    }
+  }
+
+  /**
+   * Processes products variation meta.
+   *
+   * @implements woocommerce_save_product_variation
+   */
+  public static function woocommerce_save_product_variation($postId, $loop) {
+    if (!isset($_POST['variable_post_id'])) {
+      return;
+    }
+    $variationId = $_POST['variable_post_id'][$loop];
+
+    if ($warehouses = Config::get('warehouses')) {
+      static::saveProductVariationsCustomFields($warehouses, 'warehouse', $loop, $variationId);
+    }
+
+    if ($showrooms = Config::get('showrooms')) {
+      static::saveProductVariationsCustomFields($showrooms, 'showroom', $loop, $variationId);
+    }
+  }
+
+  /**
+   * Saves custom fields for products variations.
+   *
+   * @param array $fields
+   *   Fields to render.
+   * @param string $fieldsIdPrefix
+   *   Prefix used to identify the fields.
+   * @param int $loop
+   *   The position in variations loop.
+   * @param int $variationId
+   *   The ID of the product variation.
+   */
+  protected static function saveProductVariationsCustomFields($fields, $fieldsIdPrefix, $loop, $variationId) {
+    foreach ($fields as $field) {
+      $field = sprintf('_%s_%s_%s', Plugin::PREFIX, $fieldsIdPrefix, $field['id']);
```

Comment:
The plugin should provide CRUD API methods so that @marcbinz doesn't have to build this very custom meta key name from scratch on his own.

Even better would be a Stock class:
```php
$stock = new Stock($product);
$stock->setValue('A', $value);
$stock->setValue('B', $value);
...
```
And/or a Store:
```php
$store = new Store('A');
$store->setStock($product, $value);
$store = new Store('B');
$store->setStock($product, $value);
...
```
@marcbinz thinks that the Store class will probably more useful for the Moeve import/sync cases.